src/main/kotlin/com/manimdsl/frontend/SymbolTable.kt

```diff
@@ -2,30 +2,51 @@ package com.manimdsl.frontend
 
 data class IdentifierData(val type: Type)
 
-interface SymbolTable {
-    fun getTypeOf(identifier: String): Type
-}
+class SymbolTable {
+    private val scopes = mutableListOf<ScopedSymbolTable>(GlobalScopeSymbolTable())
+    private var currentScope: ScopedSymbolTable = scopes[0]
+
+    fun getTypeOf(identifier: String): Type = currentScope.getTypeOf(identifier)
+
+    fun addVariable(identifier: String, type: Type) {
+        currentScope.addVariable(identifier, type)
+    }
+
+    fun enterScope(): Int {
+        val newScope = ScopedSymbolTable(currentScope)
+        scopes.add(newScope)
+        currentScope = newScope
+        // Returning index of this new symbol table (used in AST)
+        return scopes.size - 1
+    }
+
+    fun leaveScope() {
+        currentScope = currentScope.parent?: scopes[0]
+    }
 
-object SymbolTableRoot : SymbolTable {
-    override fun getTypeOf(identifier: String): Type {
-        return NoType
+    fun goToScope(scopeID: Int) {
+        if (scopeID in scopes.indices) {
+            currentScope = scopes[scopeID]
+        }
     }
+
+    fun getCurrentScopeID(): Int = scopes.indexOf(currentScope)
 }
 
-class SymbolTableNode : SymbolTable {
+class GlobalScopeSymbolTable: ScopedSymbolTable(null)
+
+open class ScopedSymbolTable(val parent: ScopedSymbolTable?) {
```

Comment:
I don't understand why there are so many SymbolTable "versions"? Is it not just a tree of SymbolTables with the root being the GlobalSymbolTable?

Why do we need ScopedSymbolTable and SymbolTable?

Reply:
I think we may need to refactor actually. Give us a bit

Reply:
This is because the SymbolTable is keeping track of the whole tree and allowing jumps to scopes which otherwise aren't accessible to the currentScope

Reply:
Why do we need jumps to scopes? Don't we just need pointers upwards and traversing upwards for the search?

Reply:
In the ASTExecutor we will be traversing the AST so the AST nodes will have the scope ids

Reply:
That still doesn't make sense to me, you don't need to be jumping you just traverse down

Reply:
The whole point of a tree is that scoping is for the children, so make a doubly linked tree and then you're done? Go up for searching if a variable is in scope, and go down when visiting the AST nodes in the executor using enter scope and a accumalator?